File: challenges/499/284.PeekingIterator.go
package challenges

/*
Given an Iterator class interface with methods: next() and hasNext(), design and implement a PeekingIterator that support the peek() operation -- it essentially peek() at the element that will be returned by the next call to next().

Example:

Assume that the iterator is initialized to the beginning of the list: [1,2,3].

Call next() gets you 1, the first element in the list.
Now you call peek() and it returns 2, the next element. Calling next() after that still return 2.
You call next() the final time and it returns 3, the last element.
Calling hasNext() after that should return false.
Follow up: How would you extend your design to be generic and work with all types, not just integer?
*/

//Below is the interface for Iterator, which is already defined for you.

// Iterator ...
type Iterator struct {
	nextPtr *Iterator
	val     int
}

func (t *Iterator) hasNext() bool {
	return t.nextPtr != nil
}

func (t *Iterator) next() int {
	if t.nextPtr == nil {
		return -1
	}

	val := t.val
	t.nextPtr = t.nextPtr.nextPtr

	return val
}

// PeekingIterator ...
type PeekingIterator struct {
	cursor     *Iterator
	nextQueued bool
	nextVal    int
}

// Constructor ...
func ConstructorPI(iter *Iterator) *PeekingIterator {
	return &PeekingIterator{
		cursor:     iter,
		nextQueued: false,
		nextVal:    0,
	}
}

// hasNext ...
func (t *PeekingIterator) hasNext() bool {
	if t.nextQueued {
		return true
	}

	return t.cursor.hasNext()
}

// next ...
func (t *PeekingIterator) next() int {
	if t.nextQueued {
		val := t.nextVal
		t.nextQueued = false
		t.nextVal = -1
		return val
	}

	return t.cursor.next()
}

// peek ...
func (t *PeekingIterator) peek() int {
	if t.nextQueued {
		return t.nextVal
	}

	t.nextQueued = true
	t.nextVal = t.cursor.next()

	return t.nextVal
}

File: challenges/499/284_PeekingIterator.py
'''
Design an iterator that supports the peek operation on an existing iterator in addition to the hasNext and the next operations.

Implement the PeekingIterator class:

PeekingIterator(Iterator<int> nums) Initializes the object with the given integer iterator iterator.
int next() Returns the next element in the array and moves the pointer to the next element.
boolean hasNext() Returns true if there are still elements in the array.
int peek() Returns the next element in the array without moving the pointer.
Note: Each language may have a different implementation of the constructor and Iterator, but they all support the int next() and boolean hasNext() functions.

Example 1:

Input
["PeekingIterator", "next", "peek", "next", "next", "hasNext"]
[[[1, 2, 3]], [], [], [], [], []]
Output
[null, 1, 2, 2, 3, false]

Explanation
PeekingIterator peekingIterator = new PeekingIterator([1, 2, 3]); // [1,2,3]
peekingIterator.next();    // return 1, the pointer moves to the next element [1,2,3].
peekingIterator.peek();    // return 2, the pointer does not move [1,2,3].
peekingIterator.next();    // return 2, the pointer moves to the next element [1,2,3]
peekingIterator.next();    // return 3, the pointer moves to the next element [1,2,3]
peekingIterator.hasNext(); // return False
 

Constraints:

1 <= nums.length <= 1000
1 <= nums[i] <= 1000
All the calls to next and peek are valid.
At most 1000 calls will be made to next, hasNext, and peek.
 

Follow up: How would you extend your design to be generic and work with all types, not just integer?
'''


# Below is the interface for Iterator, which is already defined for you.
class Iterator:
    def __init__(self, nums):
        """
        Initializes an iterator object to the beginning of a list.
        :type nums: List[int]
        """

    def hasNext(self):
        """
        Returns true if the iteration has more elements.
        :rtype: bool
        """

    def next(self):
        """
        Returns the next element in the iteration.
        :rtype: int
        """


class PeekingIterator:
  def __init__(self, iterator: Iterator):
    """
    Initialize your data structure here.
    :type iterator: Iterator
    """
    self.iter = iterator
    if self.iter.hasNext():
      self.nxt_val = self.iter.next()
    else:
      self.nxt_val = 0


  def peek(self):
    """
    Returns the next element in the iteration without advancing the iterator.
    :rtype: int
    """
    
    return self.nxt_val


  def next(self):
    """
    :rtype: int
    """
    val = self.nxt_val
    if self.iter.hasNext():
      self.nxt_val = self.iter.next()
    else:
      self.nxt_val = 0
    
    return val


  def hasNext(self):
    """
    :rtype: bool
    """
    return self.nxt_val > 0
        

# Your PeekingIterator object will be instantiated and called as such:
# iter = PeekingIterator(Iterator(nums))
# while iter.hasNext():
#     val = iter.peek()   # Get the next element but not advance the iterator.
#     iter.next()         # Should return the same value as [val].